teps/0075-object-param-and-result-types.md

```diff
@@ -0,0 +1,924 @@
+      to specify a default value to use in that case instead)
+
+## Proposal
+
+* We would add support for object types for results and params, in addition to the existing string and array support
+    * Initially we would only support string values, eventually we can expand this to all values (string, array, object)
+      (note that this is the case for arrays as well, we don't yet support arrays of arrays)
+    * Only string key types would be supported (which is
+      [the only key type supported by json](https://datatracker.ietf.org/doc/html/rfc7159#section-4))
+* We would use [json object schema syntax](https://json-schema.org/understanding-json-schema/reference/object.html)
+  ([as previously suggested](https://github.com/tektoncd/pipeline/issues/1393#issuecomment-558833526) - and see also
+  [why JSON Schema](#why-json-schema)) to express the object structure
+* To support object results, we would support writing json results to the results file, as described in
+  [TEP-0076 Array results and indexing](https://github.com/tektoncd/community/pull/477)
+  (see "why json" in that proposal)
+
+This feature would be considered alpha and would be
+[(optional)gated by the alpha flag](https://github.com/tektoncd/pipeline/blob/main/docs/install.md#alpha-features).
+
+* [Defaulting to string types for values](#defaulting-to-string-types-for-values)
+* [Adding additional JSON Schema properties](#adding-additional-json-schema-properties)
+* [Why JSON Schema?](#why-json-schema)
+* [Matching interfaces and extra keys](#matching-interfaces-and-extra-keys)
+* [Required vs optional keys](#required-vs-optional-keys)
+
+### Defaulting to string types for values
+
+As an optimization, we'd support defaulting to string types for keys without the Task author needing to explicitly
+specify this. For example this more verbose specification:
+
+```yaml
+ - name: sdist
+   description: |
+     The source distribution
+      * sha: The sha256 of the contents
+      * path: Path to the resulting tar.gz
+   type: object
+   properties:
+     sha:
+       type: string
+     path:
+       type: string
+```
+
+Would be equivalent to:
+
+```yaml
+ - name: sdist
+   description: |
+     The source distribution
+      * sha: The sha256 of the contents
+      * path: Path to the resulting tar.gz
+   type: object
+   properties:
+     sha: { } # type:string is implied
+     path: { } # type:string is implied
+```
+
+This would be supported in recognition that:
+
+a. Only string types would initially be supported b. Even when other types are supported, the most common usage of
+objects will likely use string values
+
+### Adding additional JSON Schema properties
+
+This proposal suggests adding a new `properties` section to param and result definition. If we later support more json
+schema attribute such as `additionalProperties` and `required`, we'd also support them at the same level as the
+`properties` field here. (
```

Comment:
It may be something we add on a second pass, but I really would see value in having these small schemas re-usable / sharable, being able to build a library of them - perhaps in the catalog? - that can be used across tasks and contributed to by third parties.

Reusing these schemas introduces the extra complexity of versioning them, discovering them etc but I think it's still an idea worth exploring as a follow up